src/lib/addons/slack-app.ts

```diff
@@ -17,64 +21,103 @@ interface ISlackAppAddonParameters {
 export default class SlackAppAddon extends Addon {
     private msgFormatter: FeatureEventFormatter;
 
+    private accessToken?: string;
+
     private slackClient?: WebClient;
 
+    private slackChannels?: ConversationsListResponse['channels'];
+
+    private slackChannelsCacheTimeout?: NodeJS.Timeout;
+
     constructor(args: IAddonConfig) {
         super(slackAppDefinition, args);
         this.msgFormatter = new FeatureEventFormatterMd(
             args.unleashUrl,
             LinkStyle.SLACK,
         );
+        this.startCacheInvalidation();
     }
 
     async handleEvent(
         event: IEvent,
         parameters: ISlackAppAddonParameters,
     ): Promise<void> {
-        const { accessToken } = parameters;
+        try {
+            const { accessToken } = parameters;
 
-        if (!accessToken) return;
+            if (!accessToken) return;
 
-        if (!this.slackClient) {
-            this.slackClient = new WebClient(accessToken);
-        }
+            if (!this.slackClient || this.accessToken !== accessToken) {
+                this.slackClient = new WebClient(accessToken);
+            }
+
+            if (!this.slackChannels) {
+                const slackConversationsList =
+                    await this.slackClient.conversations.list({
+                        types: 'public_channel,private_channel',
+                    });
+                this.slackChannels = slackConversationsList.channels;
+            }
+
+            const taggedChannels = this.findTaggedChannels(event);
+
+            if (this.slackChannels?.length && taggedChannels.length) {
+                const slackChannelsToPostTo = this.slackChannels.filter(
```

Comment:
I'm not that familiar with the threading model of node, but what if `this.slackChannels` are set to `undefined` after we checked the length? In Java I'd create a local variable or constant to hold the current "view" of `this.slackChannels` to make sure it doesn't change in the middle of my execution

Reply:
Fair point, this should be safer: https://github.com/Unleash/unleash/pull/4284/commits/6fdb18ef6041a18d4211c583fcd994153750d210